pjsip/src/pjsip/sip_transaction.c

```diff
@@ -3274,9 +3274,11 @@ static pj_status_t tsx_on_state_completed_uas( pjsip_transaction *tsx,
 
     } else {
 	/* Ignore request to transmit. */
-	PJ_ASSERT_RETURN(event->type == PJSIP_EVENT_TX_MSG && 
-			 event->body.tx_msg.tdata == tsx->last_tx, 
-			 PJ_EINVALIDOP);
+	if (event->type == PJSIP_EVENT_TX_MSG && 
+	    event->body.tx_msg.tdata == tsx->last_tx)
+	{ 
+	    return PJ_EINVALIDOP;
+	}
```

Comment:
ups, missing the "!".

Answering the question in the comment, after I remove the assertion, there is no crash for the case of no 100rel.

As for the above code, I propose to change it as follows:
```
	PJ_ASSERT_RETURN(event->type == PJSIP_EVENT_TX_MSG, 
			 PJ_EINVALIDOP);
	/* Ignore request to transmit a new message. */
	if (event->body.tx_msg.tdata != tsx->last_tx)
	    return PJ_EINVALIDOP;
```
So, we split the original conditions, the event type checking should still raise an assertion, but for the second condition, I believe it should return an error instead if there's a request to transmit a new message. Retransmission should be okay as before.

Reply:
> ups, missing the "!".
> 
> Answering the question in the comment, after I remove the assertion, there is no crash for the case of no 100rel.

Okay, so perhaps we don't need to fix anything in this part, as the assertion is actually acceptable (considering it is app bug to send prov answer after final one), and without assertion (release mode), the error returned does not cause crash.

> 
> As for the above code, I propose to change it as follows:
> 
> ```
> 	PJ_ASSERT_RETURN(event->type == PJSIP_EVENT_TX_MSG, 
> 			 PJ_EINVALIDOP);
> 	/* Ignore request to transmit a new message. */
> 	if (event->body.tx_msg.tdata != tsx->last_tx)
> 	    return PJ_EINVALIDOP;
> ```
> 
> So, we split the original conditions, the event type checking should still raise an assertion, but for the second condition, I believe it should return an error instead if there's a request to transmit a new message. Retransmission should be okay as before.

in case modification is still needed, the proposed change is rather major as it semantically limits event type to only TX_MSG (note that theoretically there can be other event types that may get into that block).

Reply:
It is actually less restrictive than the previous one.
Originally, the assertion is:
```
pj_assert(event->type == PJSIP_EVENT_TX_MSG && 
			 event->body.tx_msg.tdata == tsx->last_tx, 
			 PJ_EINVALIDOP);
```
i.e. if it reaches the block, the event must be TX_MSG *and* a retransmission, otherwise it will trigger an assertion.

Reply:
ah okay.